### DISCUSSION TITLE
考虑支持将ip添加到系统路由中吗？

### DISCUSSION THREAD

**Discussion by happy42779** [pymumu/smartdns#1005] [Ideas]
考虑支持将解析出来的ip根据配置规则添加到系统静态路由中吗？
比如：
配置为：baidu.com/pppoe, pppoe 为接口名称
系统会添加 110.242.68.66 via pppoe 这样的路由

若不考虑的话，可以听听专业人士的高见吗？

**Comment by PikuZheng** [comment]
smartdns可以将（缓存获得的）ip添加到ipset。在iptables可以使用mangle链对这些ip打路由标记，最后在路由表中对有路由标记的指定单独的路由

**Reply by happy42779** [reply]
> smartdns可以将（缓存获得的）ip添加到ipset。在iptables可以使用mangle链对这些ip打路由标记，最后在路由表中对有路由标记的指定单独的路由

是的，这个我知道。 这也是我现在用的方式。我现在用mikrotik（主），linux（xray+dns）（旁）。 dhcp下发的网关是linux，这样如果我用手机连回家如果要让手机也能畅连，就好做很多额外的策略路由。

我现在的想法是如果smartdns可以把ip加入静态路由，再通过ospf广播给主路由，这样dhcp就可以下发mikrotik为网关，通过路由来控制数据走向，这样手机连回家就可以用同一套逻辑了。当然这只是出于折腾而已:) 

如果这个对你们来说只是多余但没有意义的工作，可以请教一下在smartdns现有的功能基础上增加这个功能，可以简单赐教下如何实现吗？ 

我现在想到的是找到添加到ipset的代码，然后在后面执行ip route add 这一逻辑。对linux系统编程不太熟悉，不知道有没有类似windows上的signal的东西来实现异步操作？

或者还请大佬指条路🤝

**Comment by PikuZheng** [comment]
太好了，我也是mikrotik+openwrt，由mikrotik决定路由且负责客户端DNS；smartdns是mikrotik的上级。
对于需要特殊路由的，一部分是ip已知（比如谷歌），另一部分是ip不定（比如akamai）。IP和域名都可以直接添加到address-lists，mangle直接根据dst-address做route（或按照规范先做mark-conntion再做mark-routing）。

您说的正常情况下linux是网关，手机连接后切换到mikrotik是网关，没理解。dhcp-server不能对单个设备指定不同的网关。

**Comment by PikuZheng** [comment]
smartdns将ip计入ipset，和mikrotik通过smartdns获取ip并写入address-list，结果应该是相同的

**Comment by happy42779** [comment]
> 太好了，我也是mikrotik+openwrt，由mikrotik决定路由且负责客户端DNS；smartdns是mikrotik的上级。
> 
> 对于需要特殊路由的，一部分是ip已知（比如谷歌），另一部分是ip不定（比如akamai）。IP和域名都可以直接添加到address-lists，mangle直接根据dst-address做route（或按照规范先做mark-conntion再做mark-routing）。
> 
> 
> 
> 您说的正常情况下linux是网关，手机连接后切换到mikrotik是网关，没理解。dhcp-server不能对单个设备指定不同的网关。

手机是指我人在外面，通过内置协议连回家的情况下。

**Reply by PikuZheng** [reply]
如果是手机或电脑拨入，建议全部路由，由家里设备负责代理上网。
如果是mikrotik拨入，可以通过脚本同步address-list。ospf只能同步route，没法处理firewall

**Comment by happy42779** [comment]
> smartdns将ip计入ipset，和mikrotik通过smartdns获取ip并写入address-list，结果应该是相同的

是的，实现的方式其实还是挺多的。您的这个方案就是把原本在linux/op上做的策略路由放到了主路由ros上了。

就我之前提到的那个设想，在已有程序基础上添加相应功能方面您有什么建议吗？

**Reply by PikuZheng** [reply]
>我现在的想法是如果smartdns可以把ip加入静态路由，再通过ospf广播给主路由，这样dhcp就可以下发mikrotik为网关，通过路由来控制数据走向，这样手机连回家就可以用同一套逻辑了。

是说这个吗？觉得问题重点是如何处理手机连接。如果手机是全部路由到mikrotik，直接连接上网也通过mikrotik代理出去，那么mikrotik自己用mangel决定路由即可（无需与smartdns交互）。如果手机需要接收mikrotik的路由宣告，那么mikrotik必须将目标网段都写到route去，不能用mangle

**Reply by happy42779** [reply]
谢谢解答，我的想法本身可能并不成熟。其实我已经大概实现了想要的功能了，只是我觉得如果能构建出准确的路由，就不必用策略路由了。再请教下，如果我想自己尝试在smartdns的基础上加这个功能，您有什么建议吗？

**Reply by PikuZheng** [reply]
我是mikrotik重度用户，没考虑过如何用其他方式解决问题。不过vpn用户拨入后，也可以将vpn用户路由（全部）指向linux，由linux负责选路（linux添加到vpn用户的反向路由即可，不需要nat

**Comment by runshine** [comment]
其实就是增加一行代码，在添加ipset的同时，调用vtysh或者系统的ip命令实现静态路由的添加，和OSPF路由扩散，进而实现无感知的动态路由功能

**Comment by kulongwangzhi85** [comment]
更希望可以，对解析结果的ip地址，进行检测。检测失败后将所有地址通过ipset或nftset功能写入。
